### ISSUE TITLE
Use i64 for ndrange intermediate result computation so that it is less prone to overflowing?

### ISSUE THREAD

**Issue by yuanming-hu** [taichi-dev/taichi#5932]
**Concisely describe the proposed feature**
```python
import taichi as ti

ti.init()

@ti.kernel
def populate():
    n = 2 ** 16
    for i, j in ti.ndrange(n, n):
        if i == j:
            print(i)

populate()
```

It prints nothing, since `ti.i32(n * n) = 0` when Taichi compute some intermediate results for ndrange.

**Describe the solution you'd like (if any)**

Let's use `ti.i64` for now, and later systematically detect overflowing in debug mode (ref: https://stackoverflow.com/questions/1815367/catch-and-compute-overflow-during-multiplication-of-two-large-integers).

**Comment by bobcao3** [comment]
This is a static range case, considering N is 32 bit, the Ndrange should know whether it requires 64 bit indices or not statically. i64 is quite slow and not supported on some platforms

**Comment by yuanming-hu** [comment]
You are right, on we don't want range to have different behavior on backends without i64 support (e.g., Metal).

**Comment by lin-hitonami** [comment]
> This is a static range case, considering N is 32 bit, the Ndrange should know whether it requires 64 bit indices or not statically. i64 is quite slow and not supported on some platforms

This is not a static for. The arguments of ti.ndrange are taichi expressions. We may need to catch the overflow in the jit evaluation inside the constant folding. 

We can add a function to the LLVM runtime using `__builtin_mul_overflow` to detect overflow for backends using llvm. For backends using SPIRV, I'm not very familiar with SPIRV, but maybe we can use `OpUMulExtended` and `OpSMulExtended`?

**Comment by bobcao3** [comment]
> > This is a static range case, considering N is 32 bit, the Ndrange should know whether it requires 64 bit indices or not statically. i64 is quite slow and not supported on some platforms
> 
> This is not a static for. The arguments of ti.ndrange are taichi expressions. We may need to catch the overflow at runtime in the jit evaluation inside the constant folding.
> 
> We can add a function to the LLVM runtime using `__builtin_mul_overflow` to detect overflow for backends using llvm. For backends using SPIRV, I'm not very familiar with SPIRV, but maybe we can use `OpUMulExtended` and `OpSMulExtended`?

Runtime overflow detection seems like a waste of performance, especially when in this case, this value in here is semantically static and is constant folded to be static in the end. Observe that `n` here is with in range, it is only out of range for `n^2`, which comes as a byproduct of the NDrange.

**Comment by lin-hitonami** [comment]
In fact our constant folding pass now constructs a small kernel and runs the kernel to get the value...

**Comment by lin-hitonami** [comment]
https://github.com/taichi-dev/taichi/blob/20116c2d8f1a8aadeceaacbe68314a26aa1213de/taichi/transforms/constant_fold.cpp#L89-L111

Or should we detect the overflow on the host after the kernel gets the folded value using `lhs`, `rhs` and `ret`? Or more aggressively can we fold the multiplication of i32 and i32 on the host without constructing the kernel and running it on the device?

**Comment by lin-hitonami** [comment]
The ndrange for can also be non-static like the program below:
```
import taichi as ti

ti.init()

@ti.kernel
def foo(a: ti.i32, b: ti.i32):
    for i, j in ti.ndrange(a, b):
        print(i, j)

foo(2 ** 16, 2 ** 16)
```
Do we want to deal with this?

**Comment by yuanming-hu** [comment]
> Do we want to deal with this?

Yes, in the majority of cases we can't constant-fold the loop bounds :-) The original program in this issue was written for simplicity, otherwise, it won't be a case that can be constant folded.
